### PR TITLE
Dev

### PR DESCRIPTION
**Pull request by rezakalfane**
Refactoring of dc-demostore-integration with testing, removing demostore related code

### CODE REVIEW

File: .vscode/launch.json
Comment (rezakalfane) [inline .vscode/launch.json:1]: Git ignore / remove

File: .gitignore [hunk 1]
```diff
@@ -11,3 +11,4 @@ logs
 !.elasticbeanstalk/*.cfg.yml
 !.elasticbeanstalk/*.global.yml
 Dockerfile.old
+localStorage/*
```
Comment (rs-amp) [inline .gitignore]: What causes this to appear?

File: src/codec/codecs/codec-error.ts [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+/**
+ * Codec error info for an HTTP error.
+ */
+export interface CodecApiErrorInfo {
+	status: number;
+	message: string;
+}
+
+/**
+* Codec error info for an HTTP error.
```
Comment (rs-amp) [inline src/codec/codecs/codec-error.ts]: GraphQL specifically.

File: .vscode/launch.json [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+            "name": "Launch Program",
+            "skipFiles": [
+                "<node_internals>/**"
+            ],
+            "program": "${workspaceFolder}/dist/index.js",
+            "preLaunchTask": "tsc: build - tsconfig.json",
+            "outFiles": [
+                "${workspaceFolder}/dist/**/*.js"
+            ],
+            "runtimeExecutable": "/Users/rsimpson/.nvm/versions/node/v14.20.0/bin/node"
```
Comment (rs-amp) [inline .vscode/launch.json:19]: This still isn't in gitignore?

File: src/amplience/index.ts [hunk 4]
```diff
@@ -1,30 +1,54 @@
 import _ from 'lodash'
 import { DemoStoreConfiguration } from '../common/types'
 import { CryptKeeper } from '../common/crypt-keeper'
+import { IntegrationError } from '../common/errors'
 
+/**
+ * Get a content item from a Dynamic Content hub.
+ * @param hub Dynamic Content hub
+ * @param args ID or key of the content item
+ * @returns
```
Comment (rs-amp) [inline src/amplience/index.ts]: Returns what?

File: src/common/api.ts [hunk 5]
```diff
@@ -1,10 +1,33 @@
...
+	Category,
+	CommonArgs,
+	CustomerGroup
+} from './types'
 
-export type API = { }
+/**
+ * Common exception type, with a string message.
+ */
+export class Exception {
```
Comment (rs-amp) [inline src/common/api.ts]: What's this actually used for 🤔

File: src/common/util.ts [hunk 6]
```diff
@@ -1,12 +1,54 @@
...
\ No newline at end of file
+/**
+ * Surround elements of a comma separated list in double quotes.
+ * @param productIds Comma separated list
+ * @returns Comma separated list with quoted items
+ */
+export const quoteProductIdString = (productIds: string) =>
+	productIds.split(',').map(quote).join(',')
+
+export const flattenConfig = (params: any = undefined): any => {
```
Comment (rs-amp) [inline src/common/util.ts:48]: No docs?

File: src/common/crypt-keeper.ts [hunk 7]
```diff
@@ -2,30 +2,61 @@ import CryptoJS from 'crypto-js'
 import rot47 from 'rot47'
 import _ from 'lodash'
 
-const reverseString = str => str.split("").reverse().join("")
+/**
+ * Reverses a string
+ * @param str 
+ * @returns
```
Comment (rs-amp) [inline src/common/crypt-keeper.ts]: Return comment missing

File: src/common/rest-client.ts [hunk 8]
```diff
@@ -1,149 +1,205 @@
...
+	...APIProperties,
+	auth_url: {
+		title: 'Oauth URL',
+		type: 'string',
+		pattern: StringPatterns.httpUrl
+	}
 }
 
+/**
+ * JSON schema properties describing APIConfiguration.
```
Comment (rs-amp) [inline src/common/rest-client.ts]: It describes client-credential properties.

File: src/middleware/index.ts [hunk 9]
```diff
@@ -1,68 +1,185 @@
...
+ */
+export type CommerceOperation =
+	| 'getProduct'
+	| 'getProducts'
+	| 'getCategory'
+	| 'getCategoryTree'
+	| 'getCustomerGroups'
+	| 'getRawProducts'
+
+export interface MiddlewareError {
```
Comment (rs-amp) [inline src/middleware/index.ts:40]: Missing doc?

File: src/middleware/index.ts [hunk 10]
```diff
@@ -1,68 +1,185 @@
...
+	type: string
+	message: string,
+	info?: object
 }
 
-export const baseConfigLocator = { config_locator: process.env.NEXT_PUBLIC_DEMOSTORE_COMMERCE_LOCATOR || process.env.NEXT_PUBLIC_DEMOSTORE_CONFIG_LOCATOR || `amprsaprod:default` }
```
Comment (rs-amp) [inline src/middleware/index.ts:11]: What used this before 🤔

File: src/common/test/rest-mock.ts [hunk 11]
```diff
@@ -0,0 +1,213 @@
...
+ * @returns 
+ */
+function combineUrls(baseUrl, relativeUrl) {
+	if (!baseUrl) return relativeUrl
+
+	return relativeUrl ? baseUrl.replace(/\/+$/, '') + '/' + relativeUrl.replace(/^\/+/, '') : baseUrl
+}
+
+/**
+ * TODO
```
Comment (rs-amp) [inline src/common/test/rest-mock.ts]: There are a few TODOs in this file.

File: src/codec/codecs/pagination.ts [hunk 12]
```diff
@@ -0,0 +1,221 @@
...
+ * Get an URL with added params from the given object.
+ * @param url The original URL
+ * @param params The params to append to the URL
+ * @returns The URL with the added params
+ */
+function applyParams(url: string, params: any): string {
+	const isRelative = url.startsWith('/')
+
+	if (isRelative) {
+		// TODO: better solution?
```
Comment (rs-amp) [inline src/codec/codecs/pagination.ts:21]: Should probably think about this.

File: src/codec/codecs/pagination.ts [hunk 13]
```diff
@@ -0,0 +1,221 @@
...
+
+		if (result.length === targetMin) {
+			break
+		}
+	}
+
+	return result
+}
+
+export async function paginateCursor<T>(
```
Comment (rs-amp) [inline src/codec/codecs/pagination.ts:185]: Missing doc

Comment (rs-amp) [reply src/codec/codecs/pagination.ts:185]: Also needs tests?

File: src/codec/codecs/common.ts [hunk 14]
```diff
@@ -1,63 +1,184 @@
...
-import { Category } from "../../common/types";
+import { CONSTANTS } from '../../common/constants'
+import { ContentType, ContentTypeSchema, ValidationLevel } from 'dc-management-sdk-js'
+import _, { Dictionary } from 'lodash'
+import { Category } from '../../common/types'
+import { env } from 'process'
+import { CodecError, CodecErrorType } from './codec-error'
 
 /**
- * @deprecated The method should not be used
```
Comment (rs-amp) [inline src/codec/codecs/common.ts:8]: Removing this warning is a bit weird. Need to investigate.

File: src/codec/codecs/common.ts [hunk 15]
```diff
@@ -1,63 +1,184 @@
...
+ * @returns Found category, if present
  */
-export const findInMegaMenu = (categories: Category[], slug: string) => {
-    return flattenCategories(categories).find(category => category.slug?.toLowerCase() === slug?.toLowerCase())
+export const findInCategoryTree = (categories: Category[], slug: string) => {
+	return flattenCategories(categories).find(category => category.slug?.toLowerCase() === slug?.toLowerCase())
 }
 
 /**
- * @deprecated The method should not be used
```
Comment (rs-amp) [inline src/codec/codecs/common.ts:15]: Same here

File: src/codec/codecs/commerce/template.ts [hunk 16]
```diff
@@ -0,0 +1,119 @@
...
+} from '../../../common'
+import _ from 'lodash'
+import { CodecPropertyConfig, CommerceCodecType, CommerceCodec } from '../core'
+import { getProductsArgError } from '../common'
+
+/**
+ * Common codec configuration.
+ */
+type CodecConfig = {
+    // productURL:         StringProperty
```
Comment (rs-amp) [inline src/codec/codecs/commerce/template.ts:17]: Why's this here?

File: src/codec/codecs/commerce/template.ts [hunk 17]
```diff
@@ -0,0 +1,119 @@
...
+
+	/**
+	 * @inheritdoc
+	 */
+	get properties(): CodecConfig {
+		return {
+			// productURL: {
+			//     title: "Product file URL",
+			//     type: "string"
+			// }
```
Comment (rs-amp) [inline src/codec/codecs/commerce/template.ts:41]: Probably remove this. Is it meant to be an example field?

File: src/codec/codecs/commerce/template.ts [hunk 18]
```diff
@@ -0,0 +1,119 @@
...
+	// instance variables
+	// products: Product[]
+	// categories: Category[]
+
+	/**
+	 * @inheritdoc
+	 */
+	async init(codecType: CommerceCodecType): Promise<CommerceCodec> {
+		// this.products = await fetchFromURL(this.config.productURL, [])
+		// this.categoryTree = this.categories.filter(cat => !cat.parent)
```
Comment (rs-amp) [inline src/codec/codecs/commerce/template.ts]: At least make these examples match the example variable names.

File: src/codec/codecs/commerce/template.ts [hunk 19]
```diff
@@ -0,0 +1,119 @@
...
+	/**
+	 * @inheritdoc
+	 */
+	async getCustomerGroups(args: CommonArgs): Promise<Identifiable[]> {
+		return []
+	}
+}
+
+export default TemplateCommerceCodecType
+// registerCodec(new TemplateCommerceCodecType())
```
Comment (rs-amp) [inline src/codec/codecs/commerce/template.ts:120]: This is done in `codecs/index.ts`, a comment here might be confusing without that context.

File: src/codec/codecs/commerce/shopify/shopify.test.ts [hunk 20]
```diff
@@ -0,0 +1,248 @@
...
+		// Test
+		const result = await codec.getProduct({ id: 'ExampleID' })
+		expect(result).toEqual(exampleProduct('ExampleID'))
+		expect(requests).toEqual([
+			productRequest('ExampleID')
+		])
+	})
+
+	test('getProducts (multiple)', async () => {
+
```
Comment (rs-amp) [inline src/codec/codecs/commerce/shopify/shopify.test.ts]: Random whitespace in most of these tests

File: src/codec/codecs/commerce/shopify/mappers.ts [hunk 21]
```diff
@@ -0,0 +1,134 @@
...
+	ShopifyVariant 
+} from './types'
+import { Dictionary } from 'lodash'
+
+/**
+ * Find the first non-empty (not null or length 0) string in a list of strings.
+ * @param strings List of strings to search
+ * @returns The first non-empty string in the list
+ */
+export const firstNonEmpty = (strings: string[]) => {
```
Comment (rs-amp) [inline src/codec/codecs/commerce/shopify/mappers.ts]: Now unused, should delete.

File: src/codec/codecs/commerce/shopify/index.ts [hunk 22]
```diff
@@ -0,0 +1,314 @@
...
+
+	/** API version */
+	version: StringProperty,
+
+	/** Site identifier */
+	site_id: StringProperty
+}
+
+/**
+ * A template commerce codec type, useful as a starting point for a new integration.
```
Comment (rs-amp) [inline src/codec/codecs/commerce/shopify/index.ts]: This is the shopify codec, not a template.

File: src/codec/codecs/commerce/shopify/index.ts [hunk 23]
```diff
@@ -0,0 +1,314 @@
...
+	/**
+	 * @inheritdoc
+	 */
+	async getApi(config: CodecPropertyConfig<CodecConfig>): Promise<CommerceAPI> {
+		return await new ShopifyCommerceCodec(config).init(this)
+	}
+}
+
+/**
+ * A template commerce codec, useful as a starting point for a new integration.
```
Comment (rs-amp) [inline src/codec/codecs/commerce/shopify/index.ts]: Same for this.

File: src/codec/codecs/commerce/shopify/index.ts [hunk 24]
```diff
@@ -0,0 +1,314 @@
...
+			}
+		})
+		this.adminApiClient = axios.create({
+			baseURL: `https://${this.config.site_id}.myshopify.com/admin/api/${this.config.version}`,
+			headers: {
+				'X-Shopify-Access-Token': this.config.admin_access_token
+			}
+		})
+
+		// this.products = await fetchFromURL(this.config.productURL, [])
```
Comment (rs-amp) [inline src/codec/codecs/commerce/shopify/index.ts]: Remove these.

File: src/codec/codecs/commerce/shopify/index.ts [hunk 25]
```diff
@@ -0,0 +1,314 @@
...
+			errors
+		}
+	}
+
+	/**
+	 * Make a request to the shopify GraphQL API.
+	 * @param query The GraphQL query string
+	 * @param variables Variables to use with the GraphQL query
+	 * @param isAdmin Whether the admin credentials must be used or not
+	 * @returns
```
Comment (rs-amp) [inline src/codec/codecs/commerce/shopify/index.ts]: Returns what?

File: src/codec/codecs/commerce/sfcc/sfcc.test.ts [hunk 26]
```diff
@@ -0,0 +1,234 @@
...
+			data: sfccSearchResult(300, 200, 1, 'newarrivals-womens')
+		},
+		'https://test.sandbox.us03.dx.commercecloud.salesforce.com/s/TestSite/dw/shop/v22_4/product_search?q=Hit&start=0&count=200': {
+			data: sfccSearchResult(300, 200, 0, 'Hit')
+		},
+		'https://test.sandbox.us03.dx.commercecloud.salesforce.com/s/TestSite/dw/shop/v22_4/product_search?q=Hit&start=200&count=200': {
+			data: sfccSearchResult(300, 200, 1, 'Hit')
+		},
+		'https://test.sandbox.us03.dx.commercecloud.salesforce.com/s/-/dw/data/v22_4/sites/TestSite/customer_groups?start=0&count=1000': {
+			// TODO: needs auth
```
Comment (rs-amp) [inline src/codec/codecs/commerce/sfcc/sfcc.test.ts]: This is checked by requests equality check.

File: src/codec/codecs/commerce/sfcc/index.ts [hunk 27]
```diff
@@ -0,0 +1,323 @@
...
+	}
+
+	/**
+	 * Fetches data from the unauthenticated axios client.
+	 * @param url URL to fetch data from
+	 * @returns Response data
+	 */
+	async fetch(url: string): Promise<any> {
+		return logResponse('get', url, (await catchAxiosErrors(async () =>
+			await axios.get(url, {
```
Comment (rs-amp) [inline src/codec/codecs/commerce/sfcc/index.ts]: use `this.axiosConfig()`

File: src/codec/codecs/commerce/commercetools/index.ts [hunk 28]
```diff
@@ -0,0 +1,255 @@
...
+
+	/**
+	 * @inheritdoc
+	 */
+	async getRawProducts(args: GetProductsArgs, method = 'getRawProducts'): Promise<CTProduct[]> {
+		let products: CTProduct[] = []
+
+		if (args.productIds) {
+			const ids = args.productIds.split(',')
+			products = mapIdentifiers<CTProduct>(ids, await paginate(this.getPage(this.rest, `/product-projections/search?filter=id:${quoteProductIdString(args.productIds)}`)))
```
Comment (rs-amp) [inline src/codec/codecs/commerce/commercetools/index.ts:229]: Does this work properly when asked for `''`?

File: src/codec/codecs/commerce/bigcommerce/mappers.ts [hunk 29]
```diff
@@ -0,0 +1,90 @@
...
+ * @param product The bigcommerce product
+ * @returns The common variant
+ */
+export const mapVariant = (variant: BigCommerceVariant, product: BigCommerceProduct): Variant => {
+	return {
+		id: `${variant.id}`,
+		sku: `${variant.sku}`,
+		listPrice: formatMoneyString(variant.calculated_price, { currency: 'USD' }),
+		salePrice: formatMoneyString(variant.sale_price, { currency: 'USD' }),
+		// attributesx: variant.option_values.map(opt => ({
```
Comment (rs-amp) [inline src/codec/codecs/commerce/bigcommerce/mappers.ts:34]: Weird that this has been disabled.

File: docs/dev/unit-testing.md [hunk 30]
```diff
@@ -0,0 +1,106 @@
...
+
+## Codecs
+
+Codec unit testing is primarily done by mocking `axios` client responses, as all codec implementations should be using axios, and real API responses can be used as the testing data. Codec tests are mandatory for being included in the project, and need to cover any edge cases that come up.
+
+### dc-integration-tester and responses for unit testing
+
+If you've written a codec already, then you can check the requests being made and the responses you're getting back with the `dc-integration-tester`, a tool to let you run methods on your integration codec. Run it with the environment variable `LOG_INTEGRATION` set to `1`, and each request the codec makes will be logged, along with the response. You should pipe stdout to a file so that it's easier to search through the requests and responses. You can do that by adding `>> file.txt` to the end of the command.
+
+You can store all your configurations in the file `<home>/.amplience/integrations.json` and quickly test using `dc-integration-tester` against any of the config:
```
Comment (rs-amp) [inline docs/dev/unit-testing.md]: Maybe clarify that the following is an example?

File: docs/dev/unit-testing.md [hunk 31]
```diff
@@ -0,0 +1,106 @@
...
+            "access_token": "<storefront access token>",
+            "admin_access_token": "<admin access token>",
+            "version": "2023-01",
+            "site_id": "<site id>"
+        }
+    }
+}
+```
+
+You can also get responses by querying the API manually, which is useful if you're focusing on tests before implementation.
```
Comment (rs-amp) [inline docs/dev/unit-testing.md]: I meant via something like insomnia, it won't be related to an image of the integration tester.

File: docs/dev/import.md [hunk 32]
```diff
@@ -0,0 +1,46 @@
...
+
+## Import the npm package
+
+```
+$ npm i @amplience/dc-demostore-integration
+```
+
+## Export middleware and init
+
+When using something like Next.js, or a similar route-based framework, this should be in `/pages/api/index.js`. This will essentially add the `/api` route tou your project.
```
Comment (rs-amp) [inline docs/dev/import.md]: `to your project`

File: docs/dev/commerce-codec.md [hunk 33]
```diff
@@ -0,0 +1,222 @@
...
+-   All products do not exist:
+    -   Return a list of all undefined.
+
+#### `getProducts` (keyword)
+
+Get multiple products by keyword.
+
+##### Edge Cases
+
+-   Not really any.
```
Comment (rs-amp) [inline docs/dev/commerce-codec.md]: Must filter by keyword, even if the vendor doesn't support it.

File: docs/dev/commerce-codec.md [hunk 34]
```diff
@@ -0,0 +1,222 @@
...
+    -   Throw a CodecError (NotFound)
+
+#### `getCategoryTree`
+
+Get a list of root level categories.
+Sub-categories should be contained in children as a tree, should not appear at root level.
+
+##### Edge Cases
+
+-   No root categories is a weird one, right now it throws but I guess it is possible for there to be none. There could be some benefit in throwing as it generally isn't meant to be empty.
```
Comment (rs-amp) [inline docs/dev/commerce-codec.md]: Decide what to do here.

File: docs/dev/add-integration.md [hunk 35]
```diff
@@ -0,0 +1,61 @@
...
+
+- `axios`: A library for performing HTTP requests. All existing test framework code expects codecs to use axios to access their APIs, so this should be used when doing any HTTP requests for consistency.
+- `OAuthRestClient`: An axios powered client that needs to authenticate with OAuth and a client+secret pair. You give this client the credentials, the api base url, and the auth URL. On the first request (and on future requests if the token has expired) it will automatically perform authentication before sending your request. If authentication fails, it will throw a CodecError.
+- `paginate`: A method that allows you to get a list of results back from a paginated endpoint, with the requested offset and size. This exists in both page number and cursor variants, which may be useful for different types of API. It is also provided with a method which is used to get a page, which allows it to be used for many different ways of doing pagination. If no offset and size are provided, the method will attempt to fetch all pages.
+  - `getPageByQuery`:
```
Comment (rs-amp) [inline docs/dev/add-integration.md]: Missing text

File: README.md [hunk 36]
```diff
@@ -1,151 +1,120 @@
...
+-   `getCategory` (by ID or slug)
+-   `getCategoryTree` (category structure)
+-   `getCustomerGroups` (customer segmentation)
 
-Takes its product catalog, category structure, and translation data from URLs specified in its configuration.
+Concrete implementations of this interface are referred to as `Codec`s and are located in `src/codec/codecs`. Platform-specific e-Commerce implementations are found in `src/codec/codecs/commerce`.
+You can find types and methods [here](./docs/dev/commerce-codec.md).
 
-#### Configuration
+There is a separate project called `dc-integration-tester` which provides a CLI for testing
```
Comment (rs-amp) [inline README.md]: Link? full stop?

**Review by rs-amp — COMMENTED** [review]
I'm also split about potentially gitignoring `dist/` and making the project build on import with a `prepare` action. It would avoid a lot of conflicts and issues with developers not rebuilding before submitting to the project for PRs.

File: src/common/api.ts (hunk 5 above)
Comment (rs-amp) [reply src/common/api.ts]: Nothing

File: .gitignore (hunk 1 above)
Comment (rezakalfane) [reply .gitignore]: It was there with earlier versions

File: .vscode/launch.json (hunk 3 above)
Comment (rezakalfane) [reply .vscode/launch.json:19]: not yet

File: src/codec/codecs/common.ts (hunk 14 above)
Comment (rs-amp) [reply src/codec/codecs/common.ts:8]: This is actively used by core codecs... Why is it different in the main branch?

File: .vscode/launch.json
Comment (rezakalfane) [reply .vscode/launch.json:1]: There is also config/settings.yaml
config/google_creds.json in .gitignore that we can remove

File: .vscode/launch.json (hunk 3 above)
Comment (rezakalfane) [reply .vscode/launch.json:19]: added

File: src/codec/codecs/commerce/commercetools/index.ts (hunk 28 above)
Comment (rezakalfane) [reply src/codec/codecs/commerce/commercetools/index.ts:229]: fix pushed, to be reviewed

File: src/middleware/index.ts (hunk 10 above)
Comment (rezakalfane) [reply src/middleware/index.ts:11]: demostore is using this I think...

File: src/codec/codecs/pagination.ts (hunk 12 above)
Comment (rezakalfane) [reply src/codec/codecs/pagination.ts:21]: maybe more something matching /^https?:\/\//i

Comment (rs-amp) [reply src/codec/codecs/pagination.ts:21]: The problem is more that it's unable to construct `URL()` with a relative url, which we use to append search params.

File: src/middleware/index.ts (hunk 10 above)
Comment (rezakalfane) [reply src/middleware/index.ts:11]: Removing all demostore related code as this will be moved to dc-demostore-core code

**Comment by rezakalfane** [comment]
> I'm also split about potentially gitignoring `dist/` and making the project build on import with a `prepare` action. It would avoid a lot of conflicts and issues with developers not rebuilding before submitting to the project for PRs.

All ok!

File: src/codec/codecs/common.ts (hunk 14 above)
Comment (rs-amp) [reply src/codec/codecs/common.ts:8]: Looking through history, dave undeprecated this in the dev branch at some point.
